### PR TITLE
Portable dialysis machine

### PR DESCRIPTION
**Pull request by stanalbatross**
## About The Pull Request
tldr
https://gitlab.com/cmdevs/colonial-warfare/-/merge_requests/2322

## Why It's Good For The Game

anti od tool for docs, synths, now that purge kits are gone

## Changelog

:cl: Stan_Albatross, Frans_Feiffer
add: Added a new medical tool - the portable dialysis machine, which is a medium sized item. Attach it to someone and it will filter out all the drugs from their bloodstream over time, but also take some of their blood in the process. Has enough charge for 100s of use, or 150u of drugs. Can be recharged. Make sure to detach it by clicking the person it's attached to with it again before walking away or putting it away, or they'll get a nasty needle wound. Sprites by Frans_Feiffer.
add: added a portable dialysis machine to lower medbay, added portable dialysis machines to the synthtool vendor for 15 points.
fix: autocompressors can now be recharged
/:cl:

### CODE REVIEW

**Comment by stanalbatross** [comment]
This was code approved on gitlab

**Review by Shad0vvs — APPROVED** [review]

File: code/game/objects/items/tools/experimental_tools.dm [hunk 1]
```diff
@@ -121,3 +121,181 @@
...
+	if(!skillcheck(user, SKILL_MEDICAL, SKILL_MEDICAL_MEDIC))
+		to_chat(user, SPAN_WARNING("You don't seem to know how to use \the [src]..."))
+		return
+
+	if(!pdcell || pdcell.charge == 0)
+		to_chat(user, SPAN_NOTICE("\The [src] flashes its 'battery low' light, and refuses to attach."))
+		return
+
+	if(ishuman(usr))
+		var/mob/living/carbon/human/H = usr
```
Comment (Watermelon914) [inline code/game/objects/items/tools/experimental_tools.dm:193]: Please don't use usr here, The attacker is `user` so use that.

File: code/game/objects/items/tools/experimental_tools.dm [hunk 2]
```diff
@@ -121,3 +121,181 @@
...
+		attached.reagents.remove_any_but(reagent_removed_per_second*delta_time, "blood")
+		attached.take_blood(attached, blood_cost*delta_time)
+		if(attached.blood_volume < BLOOD_VOLUME_SAFE) if(prob(5))
+			visible_message("\The [src] beeps loudly.")
+		pdcell.use(filter_cost*delta_time)
+
+	updateUsrDialog()
+	update_icon()
+
+/obj/item/tool/portadialysis/proc/damage_arms(var/mob/living/carbon/human/A)
```
Comment (Watermelon914) [inline code/game/objects/items/tools/experimental_tools.dm:293]: Use a more descriptive variable name here, like 'target'

File: code/modules/reagents/Chemistry-Holder.dm [hunk 3]
```diff
@@ -115,6 +115,36 @@
 	handle_reactions()
 	return total_transfered
 
+/datum/reagents/proc/remove_any_but(var/amount=1, var/reagent_to_ignore) //removes all other than the designated ignored reagent
```
Comment (Watermelon914) [inline code/modules/reagents/Chemistry-Holder.dm:118]: Default variables should not be before non-default variables.

**Review by Watermelon914 — CHANGES_REQUESTED** [review]

File: code/modules/reagents/Chemistry-Holder.dm (hunk 3 above)
Comment (stanalbatross) [reply code/modules/reagents/Chemistry-Holder.dm]: done, moved reagent to ignore first

File: code/game/objects/items/tools/experimental_tools.dm [hunk 4]
```diff
@@ -121,3 +121,181 @@
...
+	if(!skillcheck(user, SKILL_MEDICAL, SKILL_MEDICAL_MEDIC))
+		to_chat(user, SPAN_WARNING("You don't seem to know how to use \the [src]..."))
+		return
+
+	if(!pdcell || pdcell.charge == 0)
+		to_chat(user, SPAN_NOTICE("\The [src] flashes its 'battery low' light, and refuses to attach."))
+		return
+
+	if(ishuman(user))
+		var/mob/living/carbon/human/H = user
```
Comment (Watermelon914) [inline code/game/objects/items/tools/experimental_tools.dm:193]: don't need H here, user is already typecasted up to human

Comment (Watermelon914) [reply code/game/objects/items/tools/experimental_tools.dm:193]: also H isn't a detailed enough variable name anyways

Comment (stanalbatross) [reply code/game/objects/items/tools/experimental_tools.dm]: changed to user

**Comment by stanalbatross** [comment]
Sprites being sorted dnm until then

**Review by Nanu308 — APPROVED** [review]
Mapping lgtm

**Comment by stanalbatross** [comment]
just waiting on sprites

**Comment by Watermelon914** [comment]
feel free to re-open when you get the sprites for this

**Review by Agameofscones — APPROVED** [review]
maps and sprites are good, glad to see it finally done

**Review by Shad0vvs — CHANGES_REQUESTED** [review]
Fix null on if limbs are gibbed while connected and user walks away (for the human_to_damage), needs a check.
Fix making it works on rechargers.

**Review by Shad0vvs — APPROVED** [review]
